processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -925,69 +978,86 @@ long getTimestamp()
       return timestamp;
     }
 
-    String[][] getDims()
+    int[][] getDims()
     {
       return dims;
     }
 
     @Override
-    public int compareTo(TimeAndDims rhs)
+    public String toString()
     {
-      int retVal = Longs.compare(timestamp, rhs.timestamp);
-      int numComparisons = Math.min(dims.length, rhs.dims.length);
+      return "TimeAndDims{" +
+             "timestamp=" + new DateTime(timestamp) +
+             ", dims=" + Lists.transform(
+          Arrays.asList(dims), new Function<int[], Object>()
+          {
+            @Override
+            public Object apply(@Nullable int[] input)
+            {
+              if (input == null || input.length == 0) {
+                return Arrays.asList("null");
+              }
+              return Arrays.asList(input);
+            }
+          }
+      ) + '}';
+    }
+  }
+
+  Comparator<TimeAndDims> dimsComparator()
+  {
+    return new TimeAndDimsComp();
+  }
+
+  private class TimeAndDimsComp implements Comparator<TimeAndDims>
+  {
+    @Override
+    public int compare(TimeAndDims lhs, TimeAndDims rhs)
```

Comment:
this is non trivial, it will be nice to have some independent unit tests for the TimeAndDimsComp covering various corner cases.